### ISSUE TITLE
Error running sen2r() (Windows)

### ISSUE THREAD

**Issue by russmain** [ranghetti/sen2r#231]
Hi,

I'm very keen to make use of this, but have had the same issue attempting to install/run it on three different windows machines. Would very much appreciate being pointed in the right direction.
I'm new to Github, but will try my best to report according to your Issue reporting ReadME.

Installing all the dependencies works perfectly, and I get three green ticks for GDAL, aria2, and sen2cor.
However, when I run sen2r() I get the following message:

```
[2019-10-29 11:17:00] Starting sen2r execution.
GDAL version in use: 2.4.1
Error in py_module_import(module, convert = convert) : 
  ImportError: DLL load failed: The operating system cannot run %1.

Detailed traceback: 
  File "C:\OSGEO4~1\apps\Python27\lib\site-packages\gdal.py", line 2, in <module>
    from osgeo.gdal import deprecation_warn
  File "C:\OSGEO4~1\apps\Python27\lib\site-packages\osgeo\__init__.py", line 21, in <module>
    _gdal = swig_import_helper()
  File "C:\OSGEO4~1\apps\Python27\lib\site-packages\osgeo\__init__.py", line 17, in swig_import_helper
    _mod = imp.load_module('_gdal', fp, pathname, description)
In addition: Warning message:
In normalizePath(path.expand(path), winslash, mustWork) :
  path[1]="": The filename, directory name, or volume label syntax is incorrect
```
**The session info is the following:**
```
> sessionInfo()
R version 3.5.1 (2018-07-02)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 7 x64 (build 7601) Service Pack 1

Matrix products: default

locale:
[1] LC_COLLATE=English_South Africa.1252  LC_CTYPE=English_South Africa.1252   
[3] LC_MONETARY=English_South Africa.1252 LC_NUMERIC=C                         
[5] LC_TIME=English_South Africa.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] shiny_1.4.0 sen2r_1.1.0

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.2           lattice_0.20-35      class_7.3-14         shinyWidgets_0.4.9  
 [5] digest_0.6.21        foreach_1.4.7        V8_2.3               mime_0.7            
 [9] R6_2.4.0             leaflet.extras_1.0.0 e1071_1.7-2          httr_1.4.1          
[13] geojson_0.3.2        pillar_1.4.2         rlang_0.4.0          lazyeval_0.2.2      
[17] curl_4.2             rstudioapi_0.10      data.table_1.12.6    raster_3.0-7        
[21] geojsonio_0.7.0      Matrix_1.2-14        reticulate_1.13      shinyjs_1.0         
[25] rgdal_1.4-6          jqr_1.1.0            foreign_0.8-70       htmlwidgets_1.5.1   
[29] munsell_0.5.0        compiler_3.5.1       httpuv_1.5.2         pkgconfig_2.0.3     
[33] rgeos_0.5-2          htmltools_0.4.0      mapedit_0.5.0        tibble_2.1.3        
[37] codetools_0.2-15     XML_3.98-1.20        crayon_1.3.4         later_1.0.0         
[41] sf_0.8-0             grid_3.5.1           lwgeom_0.1-7         jsonlite_1.6        
[45] xtable_1.8-4         DBI_1.0.0            magrittr_1.5         units_0.6-5         
[49] scales_1.0.0         KernSmooth_2.23-15   fs_1.3.1             promises_1.1.0      
[53] leaflet_2.0.2        doParallel_1.0.15    sp_1.3-1             shinyFiles_0.7.3    
[57] iterators_1.0.12     tools_3.5.1          crosstalk_1.0.0      abind_1.4-5         
[61] parallel_3.5.1       fastmap_1.0.1        colorspace_1.4-1     stars_0.4-0         
[65] maptools_0.9-8       shinydashboard_0.7.1 classInt_0.4-2
```
The following commands produce:

```
> sen2r::load_binpaths()
$`gdalinfo`
[1] "C:\\OSGEO4~1\\bin\\gdalinfo.exe"

$ogrinfo
[1] "C:\\OSGEO4~1\\bin\\ogrinfo.exe"

$gdal_translate
[1] "C:\\OSGEO4~1\\bin\\GDAL_T~1.EXE"

$gdalwarp
[1] "C:\\OSGEO4~1\\bin\\gdalwarp.exe"

$gdalbuildvrt
[1] "C:\\OSGEO4~1\\bin\\GDALBU~1.EXE"

$gdaldem
[1] "C:\\OSGEO4~1\\bin\\gdaldem.exe"

$python
[1] "C:\\OSGEO4~1\\bin\\python.exe"

$gdal_calc
[1] "C:\\OSGEO4~1\\bin\\python.exe C:\\OSGEO4~1\\bin\\GDAL_C~1.PY"

$gdal_polygonize
[1] "C:\\OSGEO4~1\\bin\\python.exe C:\\OSGEO4~1\\bin\\GDAL_P~2.PY"

$gdal_fillnodata
[1] "C:\\OSGEO4~1\\bin\\python.exe C:\\OSGEO4~1\\bin\\GDAL_F~1.PY"

$aria2c
[1] "D:\\aria2\\aria2c.exe"

$sen2cor
[1] "D:\\Sen2Cor\\SEN2CO~1.05-\\L2A_PR~1.BAT"

attr(,"path")
[1] "~/.sen2r/paths.json"
```

```
> sen2r::init_python()
Error: 'init_python' is not an exported object from 'namespace:sen2r'
```

**Comment by russmain** [comment]
Sorry...just found the other similar issues. Will attempt to follow those suggestions first.

**Comment by ranghetti** [comment]
Hi,
I think you are referring to issue #228, right? Let me know if you will not solve foollwing it.

**Comment by russmain** [comment]
Hi,

Yeah, I've tried addressing the things mentioned in issues #154 #161 and #228, with no success as of yet. Still the same python related error.

**Comment by ranghetti** [comment]
Referring to #228, have you checked, in the OSGeo4W advanced installation, that `gdal-python` is installed?

**Comment by russmain** [comment]
Hi,

Since I use OSGeo4W to update QGIS I saw that gdal-python was installed:

![image](https://user-images.githubusercontent.com/57132771/67958348-b8f7f180-fbff-11e9-9818-993b81c531b7.png)

But I just got this message when trying to reinstall it:

![image](https://user-images.githubusercontent.com/57132771/67958402-d2993900-fbff-11e9-8cb0-a8a74f378b38.png)

Is this unrelated ? Should I close and try sort that out first...

**Comment by russmain** [comment]
Hi,

Unfortunately not. I haven't been able to dedicate a lot of time in cleaning my system, but I did try completely uninstall OSGeo and then re-install, and try clean my PATH environment...but I still get that LIBEAY32.dll conflict when trying to re-install OSGeo/gdal-python (QGIS still runs though)...and sen2r doesn't run.

I managed to install and run sen2r on my personal laptop (not an ideal long term option), which doesn't have a number of the priority software (i.e. ArcGIS) that my work desktop has, so that could be a reason, but it would be difficult to pin down exactly which one is causing the conflict.

**Comment by ranghetti** [comment]
Ok, thank you

**Comment by ranghetti** [comment]
Some improvements which I did recently should have fixed this problem. 
The reason was twofold: 1. a missing 'proj.db` file in **rgdal**; 2. the path of the Python GDAL applications was not added to PATH variable in **sen2r**.

@russmain , if you want you can test if these edits can solve your error simply installing the latest **sen2r** version from GitHub:
```r
remotes::install_github("ranghetti/sen2r")
```

**Comment by russmain** [comment]
Hi,

Thanks very much for the follow up.

I uninstalled and re-installed, but faced the followed error when running sen2r()

`[2019-12-02 08:20:40] Starting sen2r execution.
GDAL version in use: 2.4.1
Error in python_config(python_version, required_module, python_version,  : 
  Error 1 occurred running C:\OSGEO4~1\apps\qgis\bin\python.exe 
In addition: Warning messages:

1: In normalizePath(path.expand(path), winslash, mustWork) :
  path[1]="": The filename, directory name, or volume label syntax is incorrect

2: In normalizePath(path.expand(path), winslash, mustWork) :
  path[1]="NA": The system cannot find the file specified

3: In normalizePath(path.expand(path), winslash, mustWork) :
  path[1]="": The filename, directory name, or volume label syntax is incorrect

4: In normalizePath(path.expand(path), winslash, mustWork) :
  path[1]="NA": The system cannot find the file specified

5: In normalizePath(path.expand(path), winslash, mustWork) :
  path[1]="NA": The system cannot find the file specified

6: In system2(command = python, args = paste0("\"", config_script,  :
  running command '"C:\OSGEO4~1\apps\qgis\bin\python.exe" "C:/Users/RMain/Documents/R/win-library/3.5/reticulate/config/config.py"' had status 1`

**Comment by ranghetti** [comment]
There is something strange in your settings: 
```
Error 1 occurred running C:\OSGEO4~1\apps\qgis\bin\python.exe
```
but, in your first message, I see
```
$python
[1] "C:\\OSGEO4~1\\bin\\python.exe"
```

Can you check in which of the two directories (`C:\OSGEO4~1\bin` or `C:\OSGEO4~1\apps\qgis\bin`) python.exe and GDAL binaries (gdalwarp.exe, gdalinfo.exe, gdal_calc.py etc.) are?

In every case, I think an R / GDAL update could help:
1. delete the file `C:\Users\<yourusername>\Documents\.sen2r\paths.json`;
2. update GDAL by running the OSGeo4W installer in Advanced mode;
3. download and install latest R 3.6.1;
4. reinstall sen2r (to manage dependencies, it is easier to first install it from CRAN -- `install.packages("sen2r")` -- and then to install the GitHub version -- `install.packages("remotes"); remotes::install_github("ranghetti/sen2r")`).

**Comment by russmain** [comment]
> Can you check in which of the two directories (C:\OSGEO4~1\bin or C:\OSGEO4~1\apps\qgis\bin) python.exe and GDAL binaries (gdalwarp.exe, gdalinfo.exe, gdal_calc.py etc.) are?

Ah you're right...i see python.exe and gdal binaries in BOTH of those folders. Should I delete one ?

**Comment by ranghetti** [comment]
> Ah you're right...i see python.exe and gdal binaries in BOTH of those folders. Should I delete one ?

I think it would be better to follow the schema wrote above, replacing the point 2 (GDAL update) with a reinstallation. To do it, delete the whole `C:\OSGEO4~1` directory (or -- safer -- rename it) and then launch the OSGeo4W installer to reinstall all. You can both use the Advanced Install and select the **gdal-python** package -- better method if you do not need QGIS -- _or_ use the simple installation and select both GDAL and QGIS.

The reason for which I also suggest to install R 3.6.1 is that, after reinstalling GDAL, you will have GDAL 3, which should be used with **rgdal** > 1.5, which is available for R >= 3.6.

**Comment by sermomon** [comment]
I am getting the same error on a Win10 computer. I'm trying to run the following code:

`write_scihub_login(username= "my_username", password= getPass::getPass())

# SEARCH CRITERIA #
###########################################################
# adquisition date
time_window <- as.Date(c("2019-03-01", "2019-03-31"))
# tile code
tile <- "30SYJ"
# max % of clouds
max_cloud <- 20
# product level
product_level <- "L2A"
# output directory
output_directory <- "D:/DL_project/data/sentinel-2"
###########################################################

list_available <- s2_list(tile= tile, time_interval= time_window, max_cloud= max_cloud)

#safe_getMetadata(list_available, "sensing_datetime")
list_safe <- list_available
safe_getMetadata(list_safe, "nameinfo")
safe_is_online(list_safe)

out <- paste0(output_directory, "/from_", time_window[1], "_to_", time_window[2], "_", toupper(months(time_window)[1]))
dir.create(out)

output_directory <- sen2r(
  gui = FALSE,
  s2tiles_selected = tile,
  timewindow = time_window,
  max_cloud_safe = max_cloud,
  online = TRUE, order_lta = TRUE,
  step_atmcorr = "l2a",
  list_prods = c("BOA","SCL"), list_indices = c("NDVI","MSAVI2"), list_rgb = c("RGB432B"),
  mask_type = "cloud_and_shadow", max_mask = 100, mask_buffer = 30,
  extent_name = "sen2r",
  path_l1c = out, path_l2a = out, path_out = out, overwrite_safe = TRUE
)
`
Then I obtain the following error:

**[2021-06-08 13:23:10] #### Starting sen2r execution. ####
[2021-06-08 13:23:11] Searching for available SAFE products on SciHub...
[2021-06-08 13:23:17] Computing output names...
Warning: package ‘sf’ was built under R version 3.6.3
Linking to GEOS 3.8.0, GDAL 3.0.4, PROJ 6.3.1
[2021-06-08 13:23:18] Processing group 1 of 7...
[2021-06-08 13:23:18] Starting to download the required level-2A SAFE products.
[2021-06-08 13:23:18] Check if products are available for download...
[2021-06-08 13:23:19] Downloading Sentinel-2 image 1 of 1
          (S2B_MSIL2A_20190309T104019_N0211_R008_T30SYJ_20190309T150811.SAFE)...
  |=======================================================================================================| 100%
[2021-06-08 13:24:10] Download of level-2A SAFE products terminated.
[2021-06-08 13:24:10] Starting to download the required level-1C SAFE products.
[2021-06-08 13:24:10] Download of level-1C SAFE products terminated.
[2021-06-08 13:24:10] Updating output names...
[2021-06-08 13:24:10] Starting to translate SAFE products in custom format.
Using UTM zone 30N.
2 output files were correctly created.
[2021-06-08 13:24:12] Starting to merge tiles by orbit.
[2021-06-08 13:24:12] Starting to edit geometry (clip, reproject, rescale).
[2021-06-08 13:24:16] Starting to apply cloud masks.
[2021-06-08 13:24:16] Masking file S2B2A_20190309_008_sen2r_BOA_10.tif...
Traceback (most recent call last):
  File "C:\OSGEO4~2\bin\GDAL_C~1.PY", line 54, in <module>
    from osgeo import gdal
  File "C:\OSGEO4~2\apps\Python27\lib\site-packages\osgeo\__init__.py", line 32, in <module>
    _gdal = swig_import_helper()
  File "C:\OSGEO4~2\apps\Python27\lib\site-packages\osgeo\__init__.py", line 31, in swig_import_helper
    return importlib.import_module('_gdal')
  File "C:\OSGEO4~2\apps\Python27\Lib\importlib\__init__.py", line 37, in import_module
    __import__(name)
ImportError: No module named _gdal
Error in normalizePath(path.expand(path), winslash, mustWork) : 
  path[1]="C:\Users\sergi\AppData\Local\Temp\RtmpuWPlMn\sen2r_f40f46294c/groupf407fd369c2/s2_mask/mask_f40d9217d.tif": El sistema no puede encontrar el archivo especificado
[2021-06-08 13:26:00] Producing required RGB images.
[2021-06-08 13:26:00] Generating image S2B2A_20190309_008_sen2r_RGB432B_10.tif...**

I have done the 4 steps that @ranghetti proposes:

1. _Delete the file_ **C:\Users\<yourusername>\Documents\.sen2r\paths.json** . However I can't find the folder **.sen2r**

2. _Update GDAL by running the OSGeo4W installer in advanced mode_. I have my OSGeo4W folder in **C:\OSGeo4W64** and my **python.exe** is located in **C:\OSGeo4W64\bin\python.exe**

3. _Download R 3.6.1_. I am currently using R 3.6.1.

`version`
               _                           
**platform       x86_64-w64-mingw32          
arch           x86_64                      
os             mingw32                     
system         x86_64, mingw32             
status                                     
major          3                           
minor          6.1                         
year           2019                        
month          07                          
day            05                          
svn rev        76782                       
language       R                           
version.string R version 3.6.1 (2019-07-05)
nickname       Action of the Toes** 

4. _Reinstall_ `sen2r`. I `installed both install.packages("sen2r")` and `remotes::install_github("ranghetti/sen2r")` and the packages in `remotes::install_github("ranghetti/sen2r")` was updated but I obtained two warnings in this step:

**Warning: cannot remove prior installation of package ‘curl’
Warning in file.copy(savedcopy, lib, recursive = TRUE) :
  problem copying C:\Users\sergi\Documents\R\win-library\3.6\00LOCK\curl\libs\x64\curl.dll to C:\Users\sergi\Documents\R\win-library\3.6\curl\libs\x64\curl.dll: Permission denied
Warning: restored ‘curl’**

**The downloaded source packages are in
	‘C:\Users\sergi\AppData\Local\Temp\RtmpmObpBh\downloaded_packages’
WARNING: Rtools is required to build R packages, but is not currently installed.**

The `Rtools` package is not available for 3.6.1 version.

`install.packages("Rtools")`

**Installing package into ‘C:/Users/sergi/Documents/R/win-library/3.6’
(as ‘lib’ is unspecified)
Warning: package ‘Rtools’ is not available (for R version 3.6.1)**


Any ideas to fix it?

Thank you.

Sergio.

**Comment by ranghetti** [comment]
@sermomon this issue is very old; in the meanwhile several changes were performed to sen2r (https://github.com/ranghetti/sen2r/releases/tag/v1.3.7, https://github.com/ranghetti/sen2r/releases/tag/v1.3.5), in particular to make runtime GDAL an optional (and unsupported) dependency. See https://sen2r.ranghetti.info/articles/installation.html#installing-other-dependencies-generally-unrequired-and-discouraged.

In your case, removing the option `mask_buffer = 30` should be sufficient to avoid calling a runtime GDAL. Try in this way, and if you would encounter errorrs please open a new issue rather than continuing this one.

If you strictly need to use this option you must be able to configure the runtime GDAL environment following the above mentioned documentation page. Notice that, starting from version 1.3.7, no assistance is provided to configure GDAL (it was too time consuming for me, and in most of the cases the assistance did not help to solve). If you will not be able to configure GDAL, you can try sen2r through Docker (https://sen2r.ranghetti.info/articles/docker.html), in which GDAL is already configured.
